### PR TITLE
[WIP] Added EditableRepository#move()

### PR DESCRIPTION
**Pull request by wouterj**
Fixes https://github.com/puli/issues/issues/42

### CODE REVIEW

File: src/FilesystemRepository.php [hunk 1]
```diff
@@ -349,6 +378,18 @@ private function removeResource($filesystemPath, &$removed)
...
+    private function moveResource($sourcePath, $targetPath, &$moved)
+    {
+        $targetPathInBaseDir = $this->baseDir.$targetPath;
+        ++$moved;
+
+        if (is_dir($sourcePath)) {
+            $moved += $this->countChildren($sourcePath);
+        }
+
+        $this->filesystem->rename($sourcePath, $targetPathInBaseDir);
```
Comment (WouterJ) [inline src/FilesystemRepository.php]: as the filesystem uses `rename()`, I believe there is no rollback feature needed here

File: src/InMemoryRepository.php [hunk 2]
```diff
@@ -255,6 +284,20 @@ private function addResource($path, Resource $resource)
...
 
+    private function moveResource(Resource $resource, $targetPath, &$moved)
+    {
+        $sourcePath = $resource->getPath();
+
+        foreach ($this->getChildIterator($resource) as $child) {
+            $this->moveResource($child, $targetPath.'/'.Path::getFilename($child->getPath()), $moved);
+        }
+
+        $this->resources[$targetPath] = $this->resources[$sourcePath];
```
Comment (WouterJ) [inline src/InMemoryRepository.php]: There is need for some validation here (if target path does not already exists) and also add rollback functionality.

**Comment by WouterJ** [comment]
@webmozart can you please review this PR and give feedback if this is what you wanted?

**Comment by tgalopin** [comment]
This seems great to me, but it's been a while since this PR was started. What do you think of this @webmozart? If you're ok with this, we should implement `move` in the path mapping repositories.

**Comment by ElectricMaxxx** [comment]
@webmozart @WouterJ something going on here? Can we rebase and merge that PR?

File: src/Api/EditableRepository.php [hunk 3]
```diff
@@ -52,6 +52,16 @@ public function add($path, $resource);
...
 
     /**
+     * Moves a resource inside the repository.
+     *
+     * @param Resource|ResourceCollection $resourcee The resource(s) to move to
+     *                                               the path
+     * @param string                      $path      The path to move the resources
+     *                                               to
+     */
+    public function move($sourcePath, $targetPath, $language = 'glob');
```
Comment (ElectricMaxxx) [inline src/Api/EditableRepository.php]: i would miss the doc block incl. description for` $language`

Comment (dantleech) [inline src/Api/EditableRepository.php]: @WouterJ @webmozart `$sourcePath` should be a query right?

File: src/FilesystemRepository.php [hunk 4]
```diff
@@ -254,6 +254,35 @@ public function remove($query, $language = 'glob')
...
+            $this->moveResource($sourcePaths[0], $targetPath, $moved);
+        } else {
+            $this->ensureDirectoryExists($targetPath);
+
+            foreach ($sourcePaths as $sourcePath) {
+                $this->moveResource($sourcePath, $targetPath.'/'.Path::getFilename($sourcePath), $moved);
+            }
+        }
+
+        return $moved;
```
Comment (dantleech) [inline src/FilesystemRepository.php]: @WouterJ should this be declared in the documentation? Is this meant to be part of the API?

**Pull request closed without merging**